File: tests/runners/csharp/tests/Test_03_nested_namespaces.cs
// Test Case 03: Nested Namespaces
// Tests deeply nested namespace structures

using System;

class Program
{
    static int passed = 0;
    static int failed = 0;

    static void Assert(bool condition, string message)
    {
        if (!condition)
        {
            Console.WriteLine($"    FAILED: {message}");
            failed++;
        }
    }

    static void TestDeeplyNestedTable()
    {
        Console.WriteLine("  Testing deeply nested table (app.data.models.User)...");

        var user = new app.data.models.User
        {
            id = 1,
            username = "testuser"
        };

        Assert(user.id == 1, "id");
        Assert(user.username == "testuser", "username");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestNestedEnum()
    {
        Console.WriteLine("  Testing nested enum (app.data.enums.Permission)...");

        var perm = app.data.enums.Permission.Admin;
        Assert((int)perm == 3, "Admin value");

        perm = app.data.enums.Permission.Read;
        Assert((int)perm == 1, "Read value");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestCrossNamespaceReference()
    {
        Console.WriteLine("  Testing cross-namespace reference (UserService)...");

        var service = new app.services.UserService
        {
            id = 1,
            target_user_id = 42,
            permission = app.data.enums.Permission.Write
        };

        Assert(service.id == 1, "id");
        Assert(service.target_user_id == 42, "target_user_id");
        Assert(service.permission == app.data.enums.Permission.Write, "permission");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestSeparateNamespace()
    {
        Console.WriteLine("  Testing separate namespace (util.Config)...");

        var config = new util.Config
        {
            key = "debug_mode",
            value = "true"
        };

        Assert(config.key == "debug_mode", "key");
        Assert(config.value == "true", "value");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestBinaryNested()
    {
        Console.WriteLine("  Testing binary serialization with nested namespaces...");

        var original = new app.services.UserService
        {
            id = 999,
            target_user_id = 42,
            permission = app.data.enums.Permission.Admin
        };

        // Serialize
        using var ms = new System.IO.MemoryStream();
        using var writer = new System.IO.BinaryWriter(ms);
        app.services.BinaryWriters.WriteUserService(writer, original);

        // Deserialize
        ms.Position = 0;
        using var reader = new System.IO.BinaryReader(ms);
        var loaded = app.services.BinaryReaders.ReadUserService(reader);

        Assert(loaded.id == original.id, "id mismatch");
        Assert(loaded.target_user_id == original.target_user_id, "target_user_id mismatch");
        Assert(loaded.permission == original.permission, "permission mismatch");

        Console.WriteLine($"    PASS (serialized {ms.Length} bytes)");
        passed++;
    }

    static void Main()
    {
        Console.WriteLine("=== Test Case 03: Nested Namespaces ===");

        TestDeeplyNestedTable();
        TestNestedEnum();
        TestCrossNamespaceReference();
        TestSeparateNamespace();
        TestBinaryNested();

        if (failed > 0)
        {
            Console.WriteLine($"=== {failed} tests failed! ===");
            Environment.Exit(1);
        }
        else
        {
            Console.WriteLine("=== All tests passed! ===");
        }
    }
}

File: tests/runners/cpp/tests/test_03_nested_namespaces.cpp
// Test Case 03: Nested Namespaces
// Tests deeply nested namespace structures

#include <iostream>
#include <cassert>

#include "schema.hpp"
#include "schema_loaders.hpp"
#include "schema_container.hpp"

void test_deeply_nested_table() {
    std::cout << "  Testing deeply nested table (app::data::models::User)..." << std::endl;

    app::data::models::User user;
    user.id = 1;
    user.username = "testuser";

    assert(user.id == 1);
    assert(user.username == "testuser");

    std::cout << "    PASS" << std::endl;
}

void test_nested_enum() {
    std::cout << "  Testing nested enum (app::data::enums::Permission)..." << std::endl;

    app::data::enums::Permission perm = app::data::enums::Permission::Admin;
    assert(static_cast<int32_t>(perm) == 3);

    perm = app::data::enums::Permission::Read;
    assert(static_cast<int32_t>(perm) == 1);

    std::cout << "    PASS" << std::endl;
}

void test_cross_namespace_reference() {
    std::cout << "  Testing cross-namespace reference (UserService)..." << std::endl;

    app::services::UserService service;
    service.id = 1;
    service.target_user_id = 42;
    service.permission = app::data::enums::Permission::Write;

    assert(service.id == 1);
    assert(service.target_user_id == 42);
    assert(service.permission == app::data::enums::Permission::Write);

    std::cout << "    PASS" << std::endl;
}

void test_separate_namespace() {
    std::cout << "  Testing separate namespace (util::Config)..." << std::endl;

    util::Config config;
    config.key = "debug_mode";
    config.value = "true";

    assert(config.key == "debug_mode");
    assert(config.value == "true");

    std::cout << "    PASS" << std::endl;
}

void test_container_deeply_nested_table() {
    std::cout << "  Testing container access for deeply nested table..." << std::endl;

    schema_container::SchemaContainer container;

    app::data::models::User user;
    user.id = 7;
    user.username = "deep_user";
    container.users.add_row(user);

    assert(container.users.count() == 1);

    auto* found = container.users.get_by_id(7);
    assert(found != nullptr);
    assert(found->username == "deep_user");

    std::cout << "    PASS" << std::endl;
}

void test_binary_nested() {
    std::cout << "  Testing binary serialization with nested namespaces..." << std::endl;

    // Test UserService binary serialization (loaders generated for this type)
    app::services::UserService original;
    original.id = 999;
    original.target_user_id = 42;
    original.permission = app::data::enums::Permission::Admin;

    // Serialize
    std::vector<uint8_t> buffer;
    {
        polygen::BinaryWriter writer(buffer);
        polygen_loaders::write_UserService(writer, original);
    }

    // Deserialize
    app::services::UserService loaded;
    {
        polygen::BinaryReader reader(buffer);
        loaded = polygen_loaders::read_UserService(reader);
    }

    assert(loaded.id == original.id);
    assert(loaded.target_user_id == original.target_user_id);
    assert(loaded.permission == original.permission);

    std::cout << "    PASS (serialized " << buffer.size() << " bytes)" << std::endl;
}

void test_binary_deeply_nested_table() {
    std::cout << "  Testing binary serialization on deeply nested table..." << std::endl;

    app::data::models::User original;
    original.id = 7;
    original.username = "deep_user";

    std::vector<uint8_t> buffer;
    {
        polygen::BinaryWriter writer(buffer);
        polygen_loaders::write_User(writer, original);
    }

    app::data::models::User loaded;
    {
        polygen::BinaryReader reader(buffer);
        loaded = polygen_loaders::read_User(reader);
    }

    assert(loaded.id == original.id);
    assert(loaded.username == original.username);

    std::cout << "    PASS (serialized " << buffer.size() << " bytes)" << std::endl;
}

int main() {
    std::cout << "=== Test Case 03: Nested Namespaces ===" << std::endl;

    test_deeply_nested_table();
    test_nested_enum();
    test_cross_namespace_reference();
    test_separate_namespace();
    test_container_deeply_nested_table();
    test_binary_nested();
    test_binary_deeply_nested_table();

    std::cout << "=== All tests passed! ===" << std::endl;
    return 0;
}
